src/ui/public/visualize/loader/visualize_data_loader.ts

```diff
@@ -0,0 +1,109 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { isEqual } from 'lodash';
+
+// @ts-ignore
+import { VisRequestHandlersRegistryProvider as RequestHandlersProvider } from '../../registry/vis_request_handlers';
+// @ts-ignore
+import { VisResponseHandlersRegistryProvider as ResponseHandlerProvider } from '../../registry/vis_response_handlers';
+
+import { IPrivate } from '../../private';
+import {
+  RequestHandler,
+  RequestHandlerDescription,
+  RequestHandlerParams,
+  ResponseHandler,
+  ResponseHandlerDescription,
+  Vis,
+} from '../../vis';
+
+// @ts-ignore No typing present
+import { isTermSizeZeroError } from '../../elasticsearch_errors';
+
+import { toastNotifications } from 'ui/notify';
+
+function getHandler<T extends RequestHandler | ResponseHandler>(
+  from: Array<{ name: string; handler: T }>,
+  type: string | T
+): T {
+  if (typeof type === 'function') {
+    return type;
+  }
+  const handlerDesc = from.find(handler => handler.name === type);
+  if (!handlerDesc) {
+    throw new Error(`Could not find handler "${type}".`);
+  }
+  return handlerDesc.handler;
+}
+
+export class VisualizeDataLoader {
+  private requestHandler: RequestHandler;
+  private responseHandler: ResponseHandler;
+
+  private visData: any;
+  private previousVisState: any;
+  private previousRequestHandlerResponse: any;
+
+  constructor(private readonly vis: Vis, Private: IPrivate) {
+    const { requestHandler, responseHandler } = vis.type;
```

Comment:
Would this be a good opportunity to add types for `VisType.requestHandler` and `VisType.responseHandler` in `vis_type.d.ts`?